### PR TITLE
mds, client: add namespace to file_layout_t (previously ceph_file_layout)

### PR DESCRIPTION
**Pull request by liewegas**

### CODE REVIEW

File: src/mds/MDSMap.h [hunk 1]
```diff
@@ -350,6 +354,27 @@ class MDSMap {
     s = failed;
   }
 
+  // features
+  uint64_t get_up_features() {
+    if (!cached_up_features) {
+      bool first = true;
+      for (std::map<mds_rank_t, mds_gid_t>::const_iterator p = up.begin();
```
Comment (gregsfortytwo) [inline src/mds/MDSMap.h]: Not necessary, but remember you can use the auto keyword for this nonsense now. :D

File: src/include/encoding.h [hunk 2]
```diff
@@ -423,6 +423,28 @@ inline void encode(const std::list<ceph::shared_ptr<T> >& ls, bufferlist& bl)
...
+    encode(n, bl);
+    for (typename std::list<ceph::shared_ptr<T> >::const_iterator p = ls.begin(); p != ls.end(); ++p) {
+      n++;
+      encode(**p, bl, features);
+    }
+    ceph_le32 en;
+    en = n;
+    bl.copy_in(pos, sizeof(en), (char*)&en);
+  } else {
+    __u32 n = (__u32)(ls.size());    // FIXME: this is slow on a list.
```
Comment (gregsfortytwo) [inline src/include/encoding.h]: It's not slow when the size is zero. ;) I think the post-counting you're doing in the previous loop is fine too — or at least it's our current pattern.

Comment (gregsfortytwo) [inline src/include/encoding.h]: Oh, I get it, you left both in. Let's trim out this second implementation, please.

Comment (liewegas) [inline src/include/encoding.h]: This is a cut&paste of the non-shared_ptr version. I'll do a follow-up 
patch that strips it out of both.

**Comment by gregsfortytwo** [comment]
This looks good, except I'm worried about our wire encoding compatibility going forward. In particular, even with @ukernel's compat patch we're now encoding stuff a lot less clearly than we were -- directly passing in utime_t for instance, instead of doing an explicit encode_timeval on it. I foresee this biting us when @adamemerson gets his clock conversion in.

Also, we need to barrier out old MDSes once we start writing new layouts to disk. I think that should go the same as we do it with OSD stuff, but I don't see the code for it here?

**Comment by liewegas** [comment]
Adam's time series already went in.  I don't think there are plans (yet) for a wholesale conversion beyond that, but either way the ceph_time and utime_t encoding matches, so this isn't an issue (or at least is no more or less an issue than the rest of the code base).

For old clients, IIRC we decided not to prevent them from mounting, but instead to prevent them from accessing directories with layouts they didn't understand.  ...and that patch seems to have disappeared from the series :/.  I'll re-add it, but this implies that we need to use a separate feature bit for the layout than the other items going into 4.5, or merge Zheng's full support immediately, or rely on sane caps to prevent confusion.  Let's discuss in standup...

**Comment by gregsfortytwo** [comment]
As discussed verbally, as long as the clients which are lying about their support for rados namespaces return (to userspace) the same error code as the MDS does it's okay. That should be the case with Zheng's patches

@liewegas, those compatset fields are only used when creating new FSes. I don't think we have anything that automatically updates the MDSMap::compat field to lock stuff out — right now we have some explicit monitor commands to enable inline file data etc. We can either do that here or do an auto-upgrade based on connected MDS features, but either one requires the MDS to block namespace updates until the map contains the feature.

**Comment by liewegas** [comment]
https://github.com/ceph/ceph/blob/master/src/mon/MDSMonitor.cc#L508
and
https://github.com/ceph/ceph/blob/master/src/mds/MDSDaemon.cc#L926

I was surprised too but the other incompat features follow the same pattern (just have to define them and put them in the default incompat set).

**Comment by liewegas** [comment]
I'll run this through the fs suite again--the conflicts with master were nontrivial.

**Comment by gregsfortytwo** [comment]
On Wed, Feb 10, 2016 at 7:50 AM, Sage Weil <notifications@github.com> wrote:
>
> https://github.com/ceph/ceph/blob/master/src/mon/MDSMonitor.cc#L508
> and
> https://github.com/ceph/ceph/blob/master/src/mds/MDSDaemon.cc#L926
>
> I was surprised too but the other incompat features follow the same pattern (just have to define them and put them in the default incompat set).

Oh, I misread what that was doing. I...hmm. We may want to change that
at some point, but not now. :)

**Comment by adamemerson** [comment]
I still have another clock change later that WILL affect the wire. Fixing the Y2.038K bug will require a versioned/feature-bit change. (As discussed on the list.)

**Comment by ukernel** [comment]
I updated the compat patch. added code that denies access to directories with pool_ns layouts.
https://github.com/ceph/ceph-client/commits/wip-fs-filelayout

full support patches are at
https://github.com/ceph/ceph-client/commits/wip-fs-filelayout2

**Comment by ukernel** [comment]
Add 3 commits https://github.com/ukernel/ceph/commits/liewegas-wip-file-layout

**Comment by gregsfortytwo** [comment]
...crap. As @ukernel indirectly points out in dec92c5c35dc8f1ddf2dc82f42bc5a6300fb51ff, we have a small data leak across namespaces with our backtraces. :(

I guess clients who are restricted to their own namespace can't read them, so maybe this isn't a problem? It's a bit awkward though.

File: src/mds/CInode.cc [hunk 3]
```diff
@@ -3102,87 +3095,80 @@ int CInode::encode_inodestat(bufferlist& bl, Session *session,
...
   
   // do we have room?
   if (max_bytes) {
-    unsigned bytes = sizeof(e);
+    unsigned bytes = 8 + 8 + 4 + 8 + 8 + sizeof(ceph_mds_reply_cap) +
+      sizeof(struct ceph_file_layout) +
+      sizeof(struct ceph_timespec) * 3 +
+      4 + 8 + 8 + 8 + 4 + 4 + 4 + 4 + 4 +
+      8 + 8 + 8 + 8 + 8 + sizeof(struct ceph_timespec) +
+      4;
```
Comment (gregsfortytwo) [inline src/mds/CInode.cc]: d0d2361e805c830c10ed9e19924e26847dd86642 says this was wrong in the first pass. Perhaps there is a better way to size it than with raw integer math. :/

**Comment by gregsfortytwo** [comment]
http://pulpito.ceph.com/gregf-2016-02-18_11:38:57-fs-greg-pr-7098---basic-mira/

Failed some tests. :( There's a dencoder test bug with layouts and the VolumeClient test is failing (perhaps the "cluster" member got renamed?)

**Comment by gregsfortytwo** [comment]
Oh, and the valgrind failures are from this branch as well; UninitCondition from one of the new structs I guess.

(Guess I should say that test run was with @ukernel's fixes branch merged into master.)

**Comment by ukernel** [comment]
the UninitCondition valgrind issue should be fixed by https://github.com/ukernel/ceph/commit/8e88769bd6ffc35f1b3a04b19bea6a65b8792c79.

the Leak_StillReachable valgrind issue in http://qa-proxy.ceph.com/teuthology/sage-2016-02-25_11:25:28-fs-wip-sage-testing2---basic-smithi/27409/teuthology.log should be unrelated.

**Comment by liewegas** [comment]
ah, just noticed @ukernel 's patches, added to branch.

**Comment by liewegas** [comment]
http://pulpito.ceph.com/sage-2016-03-02_05:22:06-fs-wip-sage-testing2---basic-mira/

note backtraces failures is because i forgot to use wip-file-layout suite branch on this run; passes the earlier ones.

**Pull request merged**